### PR TITLE
JSON Prototypeexpander...for expanding prototypes into non-prototypes.   Does not work with scripts.  Duplicates nodes.  Alternative pull request than #830

### PR DESCRIPTION
**Pull request by coderextreme**
Added to JSON Parser pull request.  Apply that JSON Parser pull request first, otherwise, this will parse JSON files in Inline.js

### CODE REVIEW

**Comment by coderextreme** [comment]
Test page:

https://coderextreme.net/X3DJSONLD/src/main/html/x3domproto.html

**Comment by coderextreme** [comment]
Depends on https://github.com/x3dom/x3dom/pull/830

**Comment by andreasplesch** [comment]
Just a note to clarify that #830 has only
https://github.com/x3dom/x3dom/pull/830/commits/986a7467915363a8ea8099023be0eefee091a6fd

and that #831 has the same commit
https://github.com/x3dom/x3dom/pull/831/commits/986a7467915363a8ea8099023be0eefee091a6fd
plus
https://github.com/x3dom/x3dom/pull/831/commits/e20fe14d821b84594ea76c9aae4724b0c142c09d

So merging #831 would make #830 obsolete.

**Comment by andreasplesch** [comment]
The prototype expander works with json objects.

File: src/nodes/Networking/Inline.js [hunk 1]
```diff
@@ -154,9 +154,23 @@ x3dom.registerNodeType(
...
+
                 var xhr = new window.XMLHttpRequest();
-                if (xhr.overrideMimeType)
-                    xhr.overrideMimeType('text/xml');   //application/xhtml+xml
+                if (this._vf.url.length && this._vf.url[0].length) {
+			if (this._vf.url[0].endsWith(".x3d")) {
+				isJSON = false;
+				if (xhr.overrideMimeType)
+				    xhr.overrideMimeType('text/xml');   //application/xhtml+xml
+			} else if (this._vf.url[0].endsWith(".json")) {
```
Comment (andreasplesch) [inline src/nodes/Networking/Inline.js]: toLowerCase

File: src/nodes/Networking/Inline.js [hunk 2]
```diff
@@ -154,9 +154,23 @@ x3dom.registerNodeType(
...
+				if (xhr.overrideMimeType)
+				    xhr.overrideMimeType('text/json');
```
Comment (andreasplesch) [inline src/nodes/Networking/Inline.js]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Complete_list_of_MIME_types suggests application/json which seems to be official mime type. Did text/json work better with xhr ?

File: src/nodes/Networking/Inline.js [hunk 3]
```diff
@@ -205,10 +219,26 @@ x3dom.registerNodeType(
 
                     var inlScene = null, newScene = null, nameSpace = null, xml = null;
 
-                    if (navigator.appName != "Microsoft Internet Explorer")
-                        xml = xhr.responseXML;
-                    else
-                        xml = new DOMParser().parseFromString(xhr.responseText, "text/xml");
+		    if (isJSON) {
+			    console.log(xhr);
```
Comment (andreasplesch) [inline src/nodes/Networking/Inline.js]: comment assuming it is fully tested

File: src/nodes/Networking/Inline.js [hunk 4]
```diff
@@ -205,10 +219,26 @@ x3dom.registerNodeType(
...
+			    try {
+				    var json = JSON.parse(xhr.response);
+				    console.log("post parse", json);
```
Comment (andreasplesch) [inline src/nodes/Networking/Inline.js]: comment

File: src/nodes/Networking/Inline.js [hunk 5]
```diff
@@ -205,10 +219,26 @@ x3dom.registerNodeType(
...
+			
+				    // json = x3dom.protoExpander.prototypeExpander(xhr.responseURL, json);
+				    console.log("return from expander", json);
```
Comment (andreasplesch) [inline src/nodes/Networking/Inline.js]: comment

File: src/nodes/Networking/Inline.js [hunk 6]
```diff
@@ -205,10 +219,26 @@ x3dom.registerNodeType(
...
+				    var parser = new x3dom.JSONParser();
+				    xml = parser.parseJavaScript(json);
+				    console.log("post parser", xml);
```
Comment (andreasplesch) [inline src/nodes/Networking/Inline.js]: comment

File: src/JSONParser.js [hunk 7]
```diff
@@ -0,0 +1,259 @@
+﻿"use strict";
```
Comment (andreasplesch) [inline src/JSONParser.js]: license header like other files
lets put it in src/util/json

File: src/JSONParser.js [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+x3dom.JSONParser.prototype.constructor = x3dom.JSONParser;
+
+	/**
+	 * Load X3D JSON into an element.
+	 * jsobj - the JavaScript object to convert to DOM.
+	 */
+x3dom.JSONParser.prototype.parseJavaScript = function(jsobj) {
+		var child = this.CreateElement('X3D');
+		this.ConvertToX3DOM(jsobj, "", child);
+		console.log(jsobj, child);
```
Comment (andreasplesch) [inline src/JSONParser.js]: comment

File: src/JSONParser.js [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+			.replace(/&gt;/g, ">")
+			.replace(/&amp;/g, "&");
+		do {
+			str = y;
+			y = str.replace(/'([^'\r\n]*)\n([^']*)'/g, "'$1\\n$2'");
+			if (str !== y) {
+				console.log("CDATA Replacing",str,"with",y);
+			}
+		} while (y != str);
+		var domParser = new DOMParser();
```
Comment (andreasplesch) [inline src/JSONParser.js]: I think you found a better way

File: src/JSONParser.js [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+		element .appendChild(cdata);
+	};
+
+	/**
+	 * convert the object at object[key] to DOM.
+	 */
+x3dom.JSONParser.prototype.ConvertObject = function(key, object, element, containerField) {
+		var child;
+		if (object !== null && typeof object[key] === 'object') {
+			if (key.substr(0,1) === '@') {
```
Comment (andreasplesch) [inline src/JSONParser.js]: use .startsWith for readability ?

File: src/JSONParser.js [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+				this.ConvertToX3DOM(object[key], key, element);
+			} else if (key.substr(0,1) === '-') {
```
Comment (andreasplesch) [inline src/JSONParser.js]: and here

File: src/PrototypeExpander.js [hunk 12]
```diff
@@ -0,0 +1,980 @@
+x3dom.PROTOS = function() {
```
Comment (andreasplesch) [inline src/PrototypeExpander.js]: license preamble
lets also put it in src/util/json

**Comment by coderextreme** [comment]
Yes, but it requires a total revamp—which is why X3DJSONLD went to 2.0.  We should test in Edge before requiring a total upgrade.  But perhaps it’s time.

John

Sent from Mail for Windows 10

**Comment by coderextreme** [comment]
Superceded by #837, with comments except for CDATA section not acted on (XML document not available).

**Pull request merged**